CMakeLists.txt

```diff
@@ -113,84 +111,73 @@ if(HAVE_BUILTIN_EXTENSION_ZSTD)
     list(APPEND builtin_objs $<TARGET_OBJECTS:wiredtiger_zstd>)
 endif()
 
-# Establish wiredtiger library target.
-add_library(wiredtiger ${link_type} ${wt_sources} ${builtin_objs})
-
 # Generate wiredtiger.h include file.
 configure_file(src/include/wiredtiger.in "include/wiredtiger.h" @ONLY)
 # Generate our wiredtiger_config.h include file.
 configure_file(cmake/configs/wiredtiger_config.h.in "config/wiredtiger_config.h" @ONLY)
-# Set our targets public and private includes.
-target_include_directories(
-    wiredtiger
-    PUBLIC ${CMAKE_CURRENT_BINARY_DIR}/include
-    PRIVATE ${CMAKE_CURRENT_BINARY_DIR}/config ${CMAKE_CURRENT_LIST_DIR}/src/include
-)
-target_compile_options(
-    wiredtiger
-    PRIVATE ${COMPILER_DIAGNOSTIC_C_FLAGS}
-)
-# Don't treat include interface directories consumed on an imported target as system
-# directories.
-set_target_properties(wiredtiger PROPERTIES NO_SYSTEM_FROM_IMPORTED TRUE)
-
-if(ENABLE_STATIC AND NOT ENABLE_PYTHON)
-    # Avoid compiling with fPIC for static builds, this can often lead to
-    # GCC disabling many optimizations e.g. inlining, possibly introducing
-    # an overhead. We additionally can't turn off fPIC if we are compiling
-    # with python support as the python API library depends on shared library
-    # support.
-    set_property(TARGET wiredtiger PROPERTY POSITION_INDEPENDENT_CODE OFF)
-else()
-    set_property(TARGET wiredtiger PROPERTY POSITION_INDEPENDENT_CODE ON)
-endif()
 
-# Ensure we link any available library dependencies to our wiredtiger target.
-if(HAVE_LIBPTHREAD)
-    target_link_libraries(wiredtiger PUBLIC ${HAVE_LIBPTHREAD})
-    if(HAVE_LIBPTHREAD_INCLUDES)
-        target_include_directories(wiredtiger PUBLIC ${HAVE_LIBPTHREAD_INCLUDES})
+# If compiling with fPIC, we can create a intermediate library of
+# position independent objects, that both the static and shared builds
+# can use. This saving on the cost of recompiling the WiredTiger sources
+# two times over.
+if(WITH_PIC OR ENABLE_SHARED)
+    add_library(wt_objs OBJECT ${wt_sources})
+    target_include_directories(wt_objs
+        PRIVATE
+        ${CMAKE_CURRENT_BINARY_DIR}/include
+        ${CMAKE_CURRENT_BINARY_DIR}/config
+        ${CMAKE_CURRENT_LIST_DIR}/src/include
+    )
+    if(ENABLE_TCMALLOC)
+        target_include_directories(wt_objs PRIVATE ${HAVE_LIBTCMALLOC_INCLUDES})
     endif()
-endif()
-
-if(HAVE_LIBRT)
-    target_link_libraries(wiredtiger PUBLIC ${HAVE_LIBRT})
-    if(HAVE_LIBRT_INCLUDES)
-        target_include_directories(wiredtiger PUBLIC ${HAVE_LIBRT_INCLUDES})
+    if(HAVE_LIBPTHREAD)
+        target_include_directories(wt_objs PRIVATE ${HAVE_LIBPTHREAD_INCLUDES})
```

Comment:
Yep, good question! 
When declaring an object library, the objects that make up that library will be compiled independently, being separate from their use in the final `wiredtiger` library. Since whenever we compile an object i.e. `gcc foo.c -o foo.o`, if that object requires non-standard includes we need to declare them for the compiler (i.e. gcc foo.c -I bar -o foo.o). Given the wiredtiger library depends on the object library,  the object library can't inherit the includes on the `wiredtiger` target, hence we need to explicitly declare them here so that objects can be successfully compiled.
It also still useful to also have these includes declared in `define_wiredtiger_library`, as it can take in plain C sources (as well as pre-compiled objects). We do this for instance when we declare a non-PIC static library. Hence we still need to declare the includes such that the sources compiled under the `add_library` can compile properly.

Reply:
If de-duplication is preferred, I can try to generalize these includes into a helper macro that can be called both here and in `define_wiredtiger_library` i.e. `define_wiredtiger_library_includes(target)`.

Reply:
Thanks for explanation! No its okay, i like it how it is now.